front/lib/api/chat.ts

```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+  const assistantConfig = cloneBaseConfig(
+    DustProdActionRegistry["chat-assistant-wfn"].config
+  );
+
+  const assistantContext = {
+    user: {
+      username: auth.user()?.username,
+      full_name: auth.user()?.name,
+    },
+    workspace: owner.name,
+    date_today: new Date().toISOString().split("T")[0],
+  };
+
+  const messages: ChatMessageType[] = [
+    {
+      sId: new_id(),
+      role: "user",
+      message: userMessage,
+    },
+  ];
+
+  // This is an helper function that adds an error message to the `messages` array and yield the
+  // proper events in case of error.
+  function* handleError(
+    errorMessage: string
+  ): Generator<ChatMessageTriggerEvent | ChatMessageCreateEvent> {
+    messages.push({
+      sId: new_id(),
+      role: "error",
+      message: errorMessage,
+    });
+
+    yield { role: "error" } as ChatMessageTriggerEvent;
+    yield {
+      message: messages[messages.length - 1],
+    } as ChatMessageCreateEvent;
+  }
+
+  // Master loop that will run until the last message is not a "retrieval" request. It will call the
+  // assistant and push to the `messages` array. The assistant either push an "assistant" message
+  // with a response or a "retrieval" message that will contain a query, if that's the case we run
+  // the retrieval and add the retrieved documents to that last message and loop back here.
+  //
+  // At any point if an error occured an "error" message is pushed to the `messages` array and we
+  // will exit from this loop.
+  while (messages[messages.length - 1].role !== "retrieval") {
+    const res = await runActionStreamed(
+      auth,
+      "chat-assistant-wfn",
+      assistantConfig,
+      [{ messages: filterMessagesForModel(messages), assistantContext }]
+    );
+
+    if (res.isErr()) {
+      yield* handleError(
+        `Chat error: [${res.error.type}] ${res.error.message}`
+      );
+      break;
+    }
+
+    const { eventStream } = res.value;
+
+    let assistantMessageTriggered = false;
+
+    for await (const event of eventStream) {
+      if (event.type === "tokens") {
+        yield { text: event.content.tokens.text } as ChatMessageTokensEvent;
+        if (!assistantMessageTriggered) {
+          assistantMessageTriggered = true;
+          yield { role: "assistant" } as ChatMessageTriggerEvent;
+        }
+      }
+
+      if (event.type === "error") {
+        yield* handleError(event.content.message);
+        break;
+      }
+
+      if (event.type === "block_execution") {
+        const e = event.content.execution[0][0];
+        if (e.error) {
+          yield* handleError(e.error);
+          break;
+        }
+
+        if (event.content.block_name === "OUTPUT" && e.value) {
+          const m = e.value as {
+            role: MessageRole;
+            message?: string;
+            query?: string;
+          };
+
+          if (m.role === "assistant") {
+            if (!assistantMessageTriggered) {
+              assistantMessageTriggered = true;
```

Comment:
optional: would suggest cutting in a few functions, since it's getting a bit harder to read with the 6 levels of control blocks. I could take care of it in a following PR

Reply:
Yes they will have to be generators functions as we're in an async generator here. The handleError style can be used here indeed